Infer psc docs types

hdgarrood opened the pull request:
WIP fix for #1834 

I'm not quite sure why this doesn't work. The error I'm getting is "Unknown type constructor Prelude.Unit", from `examples/docs/src/UTF8.purs`.

`examples/docs` is set up like a normal PureScript package, and notably, `Prelude` is in `bower_components`. So I was wondering if that might have something to do with it. However, it's failing during type checking, before we remove dependencies modules from the output, so maybe that's just a red herring. I was also wondering if I haven't set up the `CheckEnv`, or specifically, the `Environment`, correctly.

hdgarrood commented:
Just to clarify: I'm opening this now because I'm stuck. Any suggestions?

hdgarrood commented:
It turns out, at some point between sorting the modules and the desugared modules coming out of `desugarWithEnv`, they were being reversed, so that they were in a reverse topological order, which of course broke type checking. I have no idea how, so I've just reversed them back again for now. Currently investigating.

At least the tests pass now, though. :)

hdgarrood commented:
I found out where the modules are being reversed: https://github.com/purescript/purescript/blob/2fffbcb748793d98c059ab69e0b144273368c238/src/Language/PureScript/Sugar/Names.hs#L57

No idea how this isn't breaking things elsewhere, or if we should be fixing that there...?

hdgarrood commented:
Oh also, I wonder if `ParseAndDesugar` has become a bit of a misnomer, as it's doing type-checking as well now. Really it's for transforming modules into a state where they can be used by psc-docs and psc-publish. I've been unable to come up with a better name so far.

paf31 commented on src/Language/PureScript/Sugar.hs:
```diff
@@ -65,14 +66,25 @@ import Language.PureScript.Sugar.TypeDeclarations as S
+        >=> createBindingGroupsModule
+        )
+  where
+  withSecond f (x, y) = (x,) <$> f y
```
`withSecond` is `fmap . fmap` isn't it?

hdgarrood replied:
Apparently not:

```
λ: let withSecond = (fmap . fmap) :: Functor f => (a -> f b) -> (z, a) -> f (z, b)

<interactive>:12:26:
    Couldn't match type ‘a1’ with ‘(z1, a1)’
      ‘a1’ is a rigid type variable bound by
           an expression type signature:
             Functor f1 => (a1 -> f1 b1) -> (z1, a1) -> f1 (z1, b1)
           at <interactive>:12:18
    Expected type: (a1 -> f1 b1) -> a1 -> (z1, b1)
      Actual type: (a1 -> b1) -> (z1, a1) -> (z1, b1)
    In the second argument of ‘(.)’, namely ‘fmap’
    In the expression:
        (fmap . fmap) :: Functor f => (a -> f b) -> (z, a) -> f (z, b)

<interactive>:12:26:
    Couldn't match type ‘b1’ with ‘f1 b1’
      ‘b1’ is a rigid type variable bound by
           an expression type signature:
             Functor f1 => (a1 -> f1 b1) -> (z1, a1) -> f1 (z1, b1)
           at <interactive>:12:18
    Expected type: (a1 -> f1 b1) -> a1 -> (z1, b1)
      Actual type: (a1 -> b1) -> (z1, a1) -> (z1, b1)
    In the second argument of ‘(.)’, namely ‘fmap’
    In the expression:
        (fmap . fmap) :: Functor f => (a -> f b) -> (z, a) -> f (z, b)
```

I'm not a fan of using the tuple instances in cases like these anyway, I think `\f (x, y) -> (x,) <$> f y` is always going to be clearer than anything using instances like `Functor ((,) a)`.

paf31 commented:
Looks good, but I'd like to figure out the `reverse` issue. I think that it works in the regular compiler because the list of modules is always length-one now (after the parallel build changes). I think the better place to fix it, if possible, would be in `Names.hs`.

garyb commented:
:+1: for reversing after the `foldM` in names.

hdgarrood commented:
Yeah, I think @paf31 is right. Looking at `Make` I can see why it doesn't make a difference to `psc` whether or not we reverse it. It's also easy to see why the list gets reversed if you look at the definition of `updateEnv`, so I think we can consider this issue figured out. :)

I changed it to reverse inside `Sugar.Names`, rebased, and pushed, so I think this is good to go now.

garyb commented on src/Language/PureScript/Docs/Convert/Single.hs:
```diff
@@ -135,10 +135,10 @@ basicDeclaration :: String -> DeclarationInfo -> Maybe IntermediateDeclaration
 convertDeclaration :: P.Declaration -> String -> Maybe IntermediateDeclaration
-convertDeclaration (P.TypeDeclaration _ ty) title =
-  basicDeclaration title (ValueDeclaration ty)
-convertDeclaration (P.ExternDeclaration _ ty) title =
```
Does the inferred type come out as exactly same as the annotated cases? (I'm thinking in case of type variable numbering, etc.)

hdgarrood replied:
I had assumed that explicit type declarations, where given, would be left alone by the compiler, but I will check this. But I think I've found another bug in the meantime...

hdgarrood replied:
Yes, it does seem to. For example. `Prelude.unsafeCompare` comes out as `forall a. a -> a -> Ordering`.

garyb replied:
Great, I thought it probably would, just thought it was worth checking as in error messages we end up with renamed type variables.

hdgarrood commented:
I've just found a couple more bugs. Give me a day or two.

user471 commented:
Would it fix a problem when functions doesn't show up in TAGS file if they doesn't have type declaration?

hdgarrood commented:
I don't think so, tags files are handled differently.

hdgarrood commented:
I think what is happening is that some of the desugaring steps (in particular, type class desugaring) mean that what we're getting is too desugared. For instance, type class members have started appearing at the top level in the output.

What I'm going to try now is to revert `ParseAndDesugar` to what it was before, although perhaps add in the `desugarTypeDeclarationsModule` so that it's easy to tell top level value declarations which have explicit type signatures from those that don't, and then make a note of all the value declarations which are missing type signatures, and *at the end*, typecheck everything and add the missing type signatures then. So the pipeline would be like this:

* Partially desugar (as before)
* Convert each module individually (`Convert.Single`), taking explicit type signatures where they exist
* Fully desugar everything
* Typecheck everything
* Add any missing type signatures using the result of type checking
* Add any re-exported declarations

garyb commented:
Sounds like a good approach. I forgot about the sugar passes before typechecking, so yeah I can see how that output on its own wouldn't be too useful; without the prior information it'd be impossible to reconstruct some of the declarations.

hdgarrood commented:
Cool :) using the above approach, do you think the best way to extract types after type checking would be to take them from the `Environment`? That is, if I typecheck all the modules together like I did here, will the environment from the final `CheckState` have everything in it?

garyb commented:
Yep!

hdgarrood commented:
Awesome, it seems to work :D I'll clean this up a bit and push a new branch.
